// tracked-maps-and-sets/addon/-private/map.js
import { consumeKey, consumeCollection, dirtyKey, dirtyCollection } from './util';
export class TrackedMap extends Map {
    // **** KEY GETTERS ****
    get(key) {
        consumeKey(this, key);
        return super.get(key);
    }
    has(key) {
        consumeKey(this, key);
        return super.has(key);
    }
    // **** ALL GETTERS ****
    entries() {
        consumeCollection(this);
        return super.entries();
    }
    keys() {
        consumeCollection(this);
        return super.keys();
    }
    values() {
        consumeCollection(this);
        return super.values();
    }
    forEach(fn) {
        consumeCollection(this);
        super.forEach(fn);
    }
    get size() {
        consumeCollection(this);
        return super.size;
    }
    // **** KEY SETTERS ****
    set(key, value) {
        dirtyKey(this, key);
        dirtyCollection(this);
        return super.set(key, value);
    }
    delete(key) {
        dirtyKey(this, key);
        dirtyCollection(this);
        return super.delete(key);
    }
    // **** ALL SETTERS ****
    clear() {
        super.forEach((_v, k) => dirtyKey(this, k));
        dirtyCollection(this);
        return super.clear();
    }
}
if (typeof Symbol !== undefined) {
    let originalIterator = TrackedMap.prototype[Symbol.iterator];
    Object.defineProperty(TrackedMap.prototype, Symbol.iterator, {
        get() {
            consumeCollection(this);
            return originalIterator;
        }
    });
}
export class TrackedWeakMap extends WeakMap {
    get(key) {
        consumeKey(this, key);
        return super.get(key);
    }
    has(key) {
        consumeKey(this, key);
        return super.has(key);
    }
    set(key, value) {
        dirtyKey(this, key);
        return super.set(key, value);
    }
    delete(key) {
        dirtyKey(this, key);
        return super.delete(key);
    }
}
