// tracked-maps-and-sets/addon/-private/set.js
import { consumeKey, consumeCollection, dirtyKey, dirtyCollection } from './util';
export class TrackedSet extends Set {
    // **** KEY GETTERS ****
    has(value) {
        consumeKey(this, value);
        return super.has(value);
    }
    // **** ALL GETTERS ****
    entries() {
        consumeCollection(this);
        return super.entries();
    }
    keys() {
        consumeCollection(this);
        return super.keys();
    }
    values() {
        consumeCollection(this);
        return super.values();
    }
    forEach(fn) {
        consumeCollection(this);
        super.forEach(fn);
    }
    get size() {
        consumeCollection(this);
        return super.size;
    }
    // **** KEY SETTERS ****
    add(value) {
        dirtyKey(this, value);
        dirtyCollection(this);
        return super.add(value);
    }
    delete(value) {
        dirtyKey(this, value);
        dirtyCollection(this);
        return super.delete(value);
    }
    // **** ALL SETTERS ****
    clear() {
        super.forEach((_v, k) => dirtyKey(this, k));
        dirtyCollection(this);
        return super.clear();
    }
}
if (typeof Symbol !== undefined) {
    let originalIterator = TrackedSet.prototype[Symbol.iterator];
    Object.defineProperty(TrackedSet.prototype, Symbol.iterator, {
        get() {
            consumeCollection(this);
            return originalIterator;
        }
    });
}
export class TrackedWeakSet extends WeakSet {
    has(value) {
        consumeKey(this, value);
        return super.has(value);
    }
    add(value) {
        dirtyKey(this, value);
        return super.add(value);
    }
    delete(value) {
        dirtyKey(this, value);
        return super.delete(value);
    }
}
