// tracked-maps-and-sets/addon/-private/util.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { tracked } from '@glimmer/tracking';
class Tag {
    static consumeTag(tag) {
        // read the tag value
        tag.__tag_value__;
    }
    static dirtyTag(tag) {
        // write the tag value
        tag.__tag_value__ = undefined;
    }
}
__decorate([
    tracked
], Tag.prototype, "__tag_value__", void 0);
export function createTag() {
    return new Tag();
}
export const consumeTag = Tag.consumeTag;
export const dirtyTag = Tag.dirtyTag;
////////////
const COLLECTION_SYMBOL = {};
export let consumeCollection = (obj) => {
    consumeKey(obj, COLLECTION_SYMBOL);
};
export let dirtyCollection = (obj) => {
    dirtyKey(obj, COLLECTION_SYMBOL);
};
if (Ember !== undefined) {
    // eslint-disable-next-line ember/new-module-imports
    consumeCollection = (obj) => Ember.get(obj, '[]');
    // eslint-disable-next-line ember/new-module-imports
    dirtyCollection = (obj) => Ember.notifyPropertyChange(obj, '[]');
}
////////////
const OBJECT_TAGS = new WeakMap();
function getOrCreateTag(obj, key) {
    let tags = OBJECT_TAGS.get(obj);
    if (tags === undefined) {
        tags = new Map();
        OBJECT_TAGS.set(obj, tags);
    }
    let tag = tags.get(key);
    if (tag === undefined) {
        tag = new Tag();
        tags.set(key, tag);
    }
    return tag;
}
export function consumeKey(obj, key) {
    consumeTag(getOrCreateTag(obj, key));
}
export function dirtyKey(obj, key) {
    dirtyTag(getOrCreateTag(obj, key));
}
