// tracked-maps-and-sets/commonjs/-private/map.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const util_1 = require("./util");
class TrackedMap extends Map {
    // **** KEY GETTERS ****
    get(key) {
        util_1.consumeKey(this, key);
        return super.get(key);
    }
    has(key) {
        util_1.consumeKey(this, key);
        return super.has(key);
    }
    // **** ALL GETTERS ****
    entries() {
        util_1.consumeCollection(this);
        return super.entries();
    }
    keys() {
        util_1.consumeCollection(this);
        return super.keys();
    }
    values() {
        util_1.consumeCollection(this);
        return super.values();
    }
    forEach(fn) {
        util_1.consumeCollection(this);
        super.forEach(fn);
    }
    get size() {
        util_1.consumeCollection(this);
        return super.size;
    }
    // **** KEY SETTERS ****
    set(key, value) {
        util_1.dirtyKey(this, key);
        util_1.dirtyCollection(this);
        return super.set(key, value);
    }
    delete(key) {
        util_1.dirtyKey(this, key);
        util_1.dirtyCollection(this);
        return super.delete(key);
    }
    // **** ALL SETTERS ****
    clear() {
        super.forEach((_v, k) => util_1.dirtyKey(this, k));
        util_1.dirtyCollection(this);
        return super.clear();
    }
}
exports.TrackedMap = TrackedMap;
if (typeof Symbol !== undefined) {
    let originalIterator = TrackedMap.prototype[Symbol.iterator];
    Object.defineProperty(TrackedMap.prototype, Symbol.iterator, {
        get() {
            util_1.consumeCollection(this);
            return originalIterator;
        }
    });
}
class TrackedWeakMap extends WeakMap {
    get(key) {
        util_1.consumeKey(this, key);
        return super.get(key);
    }
    has(key) {
        util_1.consumeKey(this, key);
        return super.has(key);
    }
    set(key, value) {
        util_1.dirtyKey(this, key);
        return super.set(key, value);
    }
    delete(key) {
        util_1.dirtyKey(this, key);
        return super.delete(key);
    }
}
exports.TrackedWeakMap = TrackedWeakMap;
