// tracked-maps-and-sets/commonjs/-private/set.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const util_1 = require("./util");
class TrackedSet extends Set {
    // **** KEY GETTERS ****
    has(value) {
        util_1.consumeKey(this, value);
        return super.has(value);
    }
    // **** ALL GETTERS ****
    entries() {
        util_1.consumeCollection(this);
        return super.entries();
    }
    keys() {
        util_1.consumeCollection(this);
        return super.keys();
    }
    values() {
        util_1.consumeCollection(this);
        return super.values();
    }
    forEach(fn) {
        util_1.consumeCollection(this);
        super.forEach(fn);
    }
    get size() {
        util_1.consumeCollection(this);
        return super.size;
    }
    // **** KEY SETTERS ****
    add(value) {
        util_1.dirtyKey(this, value);
        util_1.dirtyCollection(this);
        return super.add(value);
    }
    delete(value) {
        util_1.dirtyKey(this, value);
        util_1.dirtyCollection(this);
        return super.delete(value);
    }
    // **** ALL SETTERS ****
    clear() {
        super.forEach((_v, k) => util_1.dirtyKey(this, k));
        util_1.dirtyCollection(this);
        return super.clear();
    }
}
exports.TrackedSet = TrackedSet;
if (typeof Symbol !== undefined) {
    let originalIterator = TrackedSet.prototype[Symbol.iterator];
    Object.defineProperty(TrackedSet.prototype, Symbol.iterator, {
        get() {
            util_1.consumeCollection(this);
            return originalIterator;
        }
    });
}
class TrackedWeakSet extends WeakSet {
    has(value) {
        util_1.consumeKey(this, value);
        return super.has(value);
    }
    add(value) {
        util_1.dirtyKey(this, value);
        return super.add(value);
    }
    delete(value) {
        util_1.dirtyKey(this, value);
        return super.delete(value);
    }
}
exports.TrackedWeakSet = TrackedWeakSet;
