// tracked-maps-and-sets/commonjs/-private/util.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
const tracking_1 = require("@glimmer/tracking");
class Tag {
    static consumeTag(tag) {
        // read the tag value
        tag.__tag_value__;
    }
    static dirtyTag(tag) {
        // write the tag value
        tag.__tag_value__ = undefined;
    }
}
__decorate([
    tracking_1.tracked
], Tag.prototype, "__tag_value__", void 0);
function createTag() {
    return new Tag();
}
exports.createTag = createTag;
exports.consumeTag = Tag.consumeTag;
exports.dirtyTag = Tag.dirtyTag;
////////////
const COLLECTION_SYMBOL = {};
exports.consumeCollection = (obj) => {
    consumeKey(obj, COLLECTION_SYMBOL);
};
exports.dirtyCollection = (obj) => {
    dirtyKey(obj, COLLECTION_SYMBOL);
};
if (Ember !== undefined) {
    // eslint-disable-next-line ember/new-module-imports
    exports.consumeCollection = (obj) => Ember.get(obj, '[]');
    // eslint-disable-next-line ember/new-module-imports
    exports.dirtyCollection = (obj) => Ember.notifyPropertyChange(obj, '[]');
}
////////////
const OBJECT_TAGS = new WeakMap();
function getOrCreateTag(obj, key) {
    let tags = OBJECT_TAGS.get(obj);
    if (tags === undefined) {
        tags = new Map();
        OBJECT_TAGS.set(obj, tags);
    }
    let tag = tags.get(key);
    if (tag === undefined) {
        tag = new Tag();
        tags.set(key, tag);
    }
    return tag;
}
function consumeKey(obj, key) {
    exports.consumeTag(getOrCreateTag(obj, key));
}
exports.consumeKey = consumeKey;
function dirtyKey(obj, key) {
    exports.dirtyTag(getOrCreateTag(obj, key));
}
exports.dirtyKey = dirtyKey;
